Add support for running E2E tests with remote Armada server

richscott opened the pull request:
These changes add support for running the armada-spark end-to-end test suite using a remote Armada server (as opposed to a `kind`-instantiated K8s cluster on the same system as the E2E test client). This has been developed and tested on a macOS system as the Spark client with an Ubuntu Linux server running Armada (on top of a `kind` k8s cluster), as
well as the reversed (armada-spark client on Linux, Armada cluster on macOS). Some changes to fix Linux-macOS command incompatbilities were also made.

See updates to the `README.md` for the steps to use this.

Change Details:
- support for using TLS certificates is added - instructions have been added to `README.md` detailing on how to get the cert files,  define `ARMADA_MASTER` in `scripts/config.sh`, etc.
- the dev-e2e.sh has been modified to check for the existence and health of the specified Armada server, but if ARMADA_MASTER is not localhost, it will not attempt to start the server via armada-operator.
- the dev-e2e.sh script now does not use Docker to run the tests - it directly uses the local clone-copy of Spark (which is typically in `.spark-3.5.5`).
- the code to acquire a fabric8 k8s client has been changed from the default client (which assumes its running inside the destination cluster), to build the client, and use the TLS cert files if specified in the init.sh/config.sh configuration settings.

richscott committed:
Start adding support for running against remote Armada cluster server.

These changes are to allow running an armada-spark client against a
remote (i.e. non-localhost) Armada cluster server.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'master' into rich/support-remote-server

richscott committed:
Add script for getting TLS cert files for remote server

Also, fixes in TestOrchestrator for running against a remote
Armada instance, and run the tests directly, instead of using
a Docker container on the client.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'master' into rich/support-remote-server

richscott committed:
Move K8sClient class from src/test into src/main

It will soon be used by ArmadaClientApplication.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'master' into rich/support-remote-server

richscott committed:
Remove debugging messages; re-enable init-cluster on e2e script.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
In E2E script, if Armada server is not localhost, don't do certain startup tasks

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Add more README.md content on setting up for using a remote Armada server.

Signed-off-by: Rich Scott <richscott@sent.com>

GeorgeJahad commented on README.md:
```diff
@@ -80,6 +80,51 @@ Run the following command to load the Armada Spark image into your local kind cl
+
+- Generate a copy of the client TLS key, cert, and CA-cert files: (1) go into the `e2e` subdirectory, and
+run `./extract-kind-cert.sh` - it will generate `client.crt`, `client.key`, and `ca.crt`, from the output
+of `kubectl config view`. These files can be left in this directory.
+
```
Should e2e/*crt be in .gitignore?

richscott replied:
I expect that most times that the certs will be run is either when someone does testing with an external Armada cluster that is not just a Kind-based cluster that the dev/user set up on their own, so they have to get the certs from the Armada cluster administrator, and may want to track those in their armada-spark repo (at least while testing) for safekeeping/diff-tracking purposes. Probably 95+% of the time that e2e will be run will be with a local Kind-based cluster started up via the armada-operator, so these won't be generated anyway.

GeorgeJahad commented on scripts/dev-e2e.sh:
```diff
@@ -140,14 +140,19 @@ init-cluster() {
 
-  TMPDIR="$scripts/.tmp" "$AOHOME/bin/tooling/kind" load docker-image "$IMAGE_NAME" --name armada 2>&1 \
-   | log_group "Loading Docker image $IMAGE_NAME into Armada cluster";
+  if [ "$ARMADA_MASTER" = "localhost" ] ; then
```
ARMADA_MASTER is never just "localhost" is it?

usually something like: "armada://localhost:30002"

richscott replied:
This conditional has changed, and now looks like:
```
if [[ "$ARMADA_MASTER" == *"//localhost"* ]] ; then
```

GeorgeJahad commented on src/test/scala/org/apache/spark/deploy/armada/e2e/TestAssertions.scala:
```diff
@@ -18,6 +18,7 @@
 package org.apache.spark.deploy.armada.e2e
 
 import io.fabric8.kubernetes.api.model.Pod
+import org.apache.spark.deploy.armada.K8sClient
```
is this used?

richscott replied:
Yes - further down:
```
   43  case class AssertionContext(
   44      namespace: String,
   45      queueName: String,
   46      testId: String,
   47      k8sClient: K8sClient,
   48      armadaClient: ArmadaClient,
   49      podSnapshot: Option[PodSnapshot] = None
   50  ) {
   ```
   and beyond - basically, it needs to get some pods by labels, plus get ingress configs.

GeorgeJahad commented on src/test/scala/org/apache/spark/deploy/armada/e2e/TestOrchestrator.scala:
```diff
@@ -475,15 +474,18 @@ class TestOrchestrator(
       lookoutUrl: String,
       pythonScript: Option[String]
   ): Seq[String] = {
+    val sparkRepoCopy = ".spark-3.5.5"
```
hardcoded?

richscott replied:
I couldn't think of a succint, solid solution for this at the time. How about I file a separate ticket to parameterize this for now? (The tests pass (!) with both the Scala 2.13/Spark 3.5.5 and Scala 2.12/Spark 3.5.3 right now, but this should be eventually fixed.)

richscott replied:
I have filed https://github.com/G-Research/gr-oss/issues/1307
and will look for other mentions of hard-coded versions in the repo to mitigate.

GeorgeJahad commented:
Running with:
```
ARMADA_MASTER=armada://localhost:30002
```
no longer seems to work:
```
scripts/submitArmadaSpark.sh 100
Submitting spark job to Armada.
/home/gbj/incoming/armada-spark/scripts/init.sh: line 67: CLIENT_CERT_FILE: unbound variable
```

richscott committed:
Quote evaluated script vars, for CI checker

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Disable building/testing for Spark 3.5.5 config for now.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Scala linter fixes; Bash lint fixes

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Add ARMADA_LOOKOUT_URL to init.sh

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Check if TLS cert vars are defined before referencing

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Conditionally add TLS cert properties to test invocation

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Dynamically extract external IP addr for TLS cert setup

Get the first external interface IP address and use in Kind
configuration for allowing external K8S/Armada API access. Add
protective quotes around TLS vars in dev-e2e.sh, per shellcheck. Use
`realpath` for getting reliable full pathnames.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Check for busybox image for init container

Use better pattern checks for verifying if Armada master is localhost;
remove quote wrapper around tls_args, so Maven doesn't error on a ""
target.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Add init container image name.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Add diagnostic search log for .spark-* directory

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
More GH Actions debugging

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
More diagnostics for finding spark source dir

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Clone the Spark repo to get bin/spark-class and jars.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Always rebuild the Spark jars before running E2E.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'master' into rich/support-remote-server

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Fix Scala formatting violations.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Remove more refs and logic for old `local://` prefix.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'master' into rich/support-remote-server

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'master' into rich/support-remote-server

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'master' into rich/support-remote-server

richscott committed:
Return to running client code in Docker image.

Also, derive K8S server address (especially the port, as `kind`
dynamically selects a random port number to expose the api-server)
from ~/.kube/config.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Remove debugging messages; set driver details for client-mode deploys

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'master' into rich/support-remote-server

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Fix scala lint error.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Use SPARK_LOCAL_IP if defined, otherwise fall back to Kind-based addr.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Fix template file server hostname logic when running on MacOS.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Attempt to extract a good default for SPARK_LOCAL_IP

In scripts/init.sh, attempt to extract the IP address of the first
non-loopback/non-virtual network interface, and set SPARK_LOCAL_IP
for that. This is required for client deployment-mode tests, and in
certain cases (e.g. the e2e tests being run in Github CI checks), there
may not be a manual override for SPARK_LOCAL_IP (or even any local
settings in scripts/config.sh, if it does not exist).

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Get list of cluster nodes after Armada is started

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'master' into rich/support-remote-server

richscott committed:
Move logging message for Kind node list

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Don't copy example kind config designed for external Armada clusters.

The dev-e2e.sh script will always only set up an Armada cluster via
armada-operator when the cluster is local(host), so do not copy the
prototype kind-config-external-access.yaml to the armada-operator repo
copy. That file is to be used just as an example for running the Armada
Operator separately on a remote cluster machine. Also, fix the omission
of the second pseudo-node (`armada-worker2`) in the example Kind config.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on src/test/scala/org/apache/spark/deploy/armada/e2e/ArmadaOperations.scala:
```diff
@@ -213,7 +213,10 @@ class ArmadaClient(armadaUrl: String = "localhost:30002") {
-    Seq(armadactlCmd) ++ subCommand.split(" ") ++ Seq("--armadaUrl", armadaUrl)
+    // armadactl command expects the server address to be of the form
+    // <hostname-or-IP>:<port> with no pseudo-protocol prefix
+    var armadactlUrl = armadaUrl.replaceFirst("^armada://", "")
```
This code has changed substantially, and now is:
```
  212    private def buildCommand(subCommand: String): Seq[String] = {
  213      val armadactlCmd = resolveArmadactlPath.getOrElse {
  214        throw new RuntimeException("armadactl not found in system properties or PATH")
  215      }
  216      // armadactl command expects the server address to be of the form
  217      // <hostname-or-IP>:<port> with no pseudo-protocol prefix
  218      val pattern      = """.*armada://(.+)""".r
  219      var armadactlUrl = "undefined-armadactl-url"
  220
  221      armadaUrl match {
  222        case pattern(hostPort) => armadactlUrl = hostPort // e.g. "localhost:30002"
  223        case _ =>
  224          throw new RuntimeException(
  225            s"could not extract valid armadactl URL from armada URL ${armadaUrl}"
  226          )
  227      }
  228
  229      Seq(armadactlCmd) ++ subCommand.split(" ") ++ Seq("--armadaUrl", armadactlUrl)
  230    }
  ```
  I declare `armadactlUrl` with `var` and set its initial value fo "undefined-armada-url", which is overwritten in a successful pattern match, so the default value will make apparent to the user that something did not get set correctly.

GeorgeJahad replied:
i don't think the user will ever see the default value, "undefined-armadactl-url" after the pattern match.  It will either match or throw.

so i think getting rid of the var and writing the val like so is more idiomatic:

```
    val armadactlUrl = armadaUrl match {
      case pattern(hostPort) => hostPort
```

richscott replied:
👍🏼   Code now changed to use this strategy in commit c5c02d5.

richscott commented on src/test/scala/org/apache/spark/deploy/armada/e2e/TestOrchestrator.scala:
```diff
@@ -262,9 +263,8 @@ class TestOrchestrator(
-    val dockerCommand = buildDockerCommand(
+    val runTestCommand = buildRunTestCommand(
       config.imageName,
-      volumeMounts,
```
That was a by-product of the attempt to de-Docker-ize the client submission logic; this line is now back in.

richscott committed:
Merge branch 'master' into rich/support-remote-server

richscott commented:
> Running with:
> 
> ```
> ARMADA_MASTER=armada://localhost:30002
> ```
> 
> no longer seems to work:
> 
> ```
> scripts/submitArmadaSpark.sh 100
> Submitting spark job to Armada.
> /home/gbj/incoming/armada-spark/scripts/init.sh: line 67: CLIENT_CERT_FILE: unbound variable
> ```

This has been fixed.  I just now verified with
```
scripts/submitArmadaSpark.sh -p --mode cluster --allocation dynamic 20
```
and it ran successfully.

richscott commented on src/test/scala/org/apache/spark/deploy/armada/e2e/TestOrchestrator.scala:
```diff
@@ -475,15 +474,18 @@ class TestOrchestrator(
+
+    val classPathEntries: Seq[String] = Seq(
+      ".",
+      s"${sparkRepoCopy}/assembly/target/scala-2.13/jars/*",
```
I'll include this in the "parameterize spark and scala versions in e2e test orchestration code" ticket, if that's okay.

sudiptob2 commented on .github/workflows/build.yaml:
```diff
@@ -13,16 +13,16 @@ jobs:
-          - scala_version: "2.12.15"
-            spark_version: "3.3.4"
-            java_version: "11"
+          # - scala_version: "2.12.15"
+          # spark_version: "3.3.4"
+          # java_version: "11"
```
Are these supposed to be commented out?

richscott replied:
Yes, this was per Dave G's instruction a couple months ago, to temporarily defer working on  support for this version, since test run-time and correctness was so problematic at the time.

sudiptob2 reviewed (commented):
Small question, will do a thorough review shortly

richscott committed:
Remove debugging print statements from K8sClient.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Remove commented-out code.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented:
> Running with:
> 
> ```
> ARMADA_MASTER=armada://localhost:30002
> ```
> 
> no longer seems to work:
> 
> ```
> scripts/submitArmadaSpark.sh 100
> Submitting spark job to Armada.
> /home/gbj/incoming/armada-spark/scripts/init.sh: line 67: CLIENT_CERT_FILE: unbound variable
> ```

That breakage has been fixed since this was observed. I am able to run `scripts/submitArmadaSpark.sh 100` without error, and I also just ran `scripts/submitArmadaSpark.sh -p --mode cluster --allocation dynamic 20` which ran successfully.

sudiptob2 commented on src/test/scala/org/apache/spark/deploy/armada/e2e/K8sClient.scala:
```diff
@@ -15,21 +15,75 @@
+  clusters.foreach { entry =>
+    val cluster = entry.get("cluster").asInstanceOf[java.util.Map[String, Object]]
+    k8sApiURL = cluster.get("server").toString()
+  }
```
If I have multiple clusters in `~/.kube/config`, it will get the last one, which might not be the actual target cluster all the time.

richscott replied:
Good catch - this logic (and the 10 lines before it) should really just be run if ARMADA_MASTER is like"armada://local....", and the cluster is a local kind cluster. Otherwise, we should derive k8sApiURL from ARMADA_MASTER. (Well, I suppose it might be possible to have a Armada cluster where the Armada API server is different than the underlying K8S apiserver, but that would be very rare and strange.) I'll change this code to do this extraction from .kube/config if it's a local cluster (started by Kind).

richscott replied:
In commit c5c02d5, it now uses only the cluster that matches the `current-context` value.

sudiptob2 commented on src/test/scala/org/apache/spark/deploy/armada/e2e/TestOrchestrator.scala:
```diff
@@ -653,9 +663,17 @@ class TestOrchestrator(
+        localIP = sys.env("SPARK_LOCAL_IP")
+      } else {
+        localIP = "172.18.0.1"
+      }
```
Probably we can use `sys.env.getOrElse`  here instead of doing if else

GeorgeJahad replied:
yup:
val localIP = sys.env.getOrElse("SPARK_LOCAL_IP", "172.18.0.1")

richscott replied:
Aha, I will try that.  

update: fixed in commit 611b5db.

sudiptob2 commented on scripts/dev-e2e.sh:
```diff
@@ -138,45 +173,85 @@ init-cluster() {
+
   # Run the Scala E2E test suite
-  mvn scalatest:test -Dsuites="org.apache.spark.deploy.armada.e2e.ArmadaSparkE2E" \
+  # env MAVEN_OPTS='-Dcom.sun.net.ssl.checkRevocation=false'
```
Should this be commented out?

richscott replied:
It should be just removed - this was used during TLS cert testing. Fixed in commit 611b5db.

sudiptob2 commented on e2e/kind-config-external-access.yaml:
```diff
@@ -0,0 +1,53 @@
+    hostPort: 30002
+    protocol: TCP
+  # Kubernetes API
+  - containerPort: 6443
+    hostPort: 6443
+    protocol: TCP
```
Question: we are exposing the KubeAPI here. If the end goal is to eventually run E2E tests against production-like Armada clusters, the Kubernetes API will likely not be exposed there. current approach relies on fabric8 and direct Kubernetes access for assertions, which may not work in that environment.

GeorgeJahad replied:
I'm not sure of the specifics here, but the general point is valid.   Our internal cluster has specific auth requirements.

Have you had a chance to consider how the k8sClient would work with those?

richscott replied:
Well, we are also already using fabric8 in parts of the non-test code in src/main/ - related to job-template loading, pod-spec handling, creating an armada job-request object from a fabric8 pod object, serialization of pods, etc. We don't use a lot of the K8S api directly, but some of it - without that available, armada-spark would not run at all. I think only a small subset of K8S API is exposed to us, in the prod scenario(s).

sudiptob2 replied:
The previous use of fabric8 in `src/main` does not seem to make any network calls. Methods like `getMetadata()`, `getLabels()`, and `getAnnotations()` on `ObjectMeta` look like simple Java getters that return private fields (for example, `return this.labels;`). There is no reference to `KubernetesClient`.

I may be missing something, but this seems worth investigating further.

richscott replied:
Update: we do use some fabric8 classes in src/main, however they are just model-type classes and serialization methods - no network transport occurs in that usage. In src/test, we do access the k8s api via fabric8, to verify pods, etc.

sudiptob2 commented on src/test/scala/org/apache/spark/deploy/armada/e2e/K8sClient.scala:
```diff
@@ -15,21 +15,75 @@
  */
 
-package org.apache.spark.deploy.armada.e2e
+package org.apache.spark.deploy.armada
```
Since this file is being used in tests only (correct me if I am wrong) , should we move it under src/test?

richscott replied:
Good suggestion - this class was originally written by another developer before me, and I didn't notice that it's only used in src/test -  I will move it.

richscott replied:
Fixed in commit 611b5db.

sudiptob2 commented on src/test/scala/org/apache/spark/deploy/armada/e2e/K8sClient.scala:
```diff
@@ -15,21 +15,75 @@
+  }
+
+  val cfg = cb
+    .withClientKeyAlgo("RSA")
```
Is it OK to hardcode RSA here? I guess there might be other algorithms that different users will use.

richscott replied:
In commit d718ab6  this is now dynamically derived from the client certificate file.

sudiptob2 commented on src/test/scala/org/apache/spark/deploy/armada/e2e/K8sClient.scala:
```diff
@@ -15,21 +15,75 @@
 
 import org.apache.spark.deploy.armada.Config
-import io.fabric8.kubernetes.client.{DefaultKubernetesClient, KubernetesClient}
+import org.apache.spark.deploy.armada.submit
```
looks unused

richscott replied:
Fixed in commit 7239321.

sudiptob2 reviewed (commented):
Thanks for the PR! I did an initial review and left a few comments.

GeorgeJahad commented on src/test/scala/org/apache/spark/deploy/armada/e2e/TestOrchestrator.scala:
```diff
@@ -74,6 +75,7 @@ class TestOrchestrator(
     armadaClient: ArmadaClient,
     k8sClient: K8sClient
 )(implicit ec: ExecutionContext) {
+  val sparkRepoCopy = ".spark-3.5.5"
```
is this used anywhere?

richscott replied:
Good catch - nope it is now not used - removed in commit 7e879dd.

GeorgeJahad commented on e2e/kind-config-external-access.yaml:
```diff
@@ -0,0 +1,53 @@
+#
+# This configuration will allow you to run kubectl and armadactl
+# against the Armada instance on this system. To use this:
+# - Copy your $HOME/.kube/config on this system to the same directory
```
This section is pretty hard to read and confusing.  Something like the following is clearer to me, (you'll have to check it for accuracy.):
```
# Setup (on each remote client host):
#   1. Copy $HOME/.kube/config from the server to $HOME/.kube/config on the
#      client. Edit the client copy: change "server: https://0.0.0.0:6443"
#      to use the server's external IP.
#   2. Copy $HOME/.armadactl.yaml from the server to $HOME/.armadactl.yaml
#      on the client. Edit the client copy:
#        - Change the 'armadaUrl' value from 'localhost' to the server's
#          external IP (or hostname).
#        - Add the following line at the same indent level:
#            forceNoTls: true
#
# Verify:
#   kubectl cluster-info
#   armadactl get queues

richscott replied:
I agree - your instructions are succint and clearer - this now used in commit 2aeee4e.

GeorgeJahad commented on scripts/init.sh:
```diff
@@ -112,9 +113,34 @@ export ARMADA_EVENT_WATCHER_USE_TLS=${ARMADA_EVENT_WATCHER_USE_TLS:-false}
 export SPARK_SECRET_KEY="${SPARK_SECRET_KEY:-armada-secret}"
 
+if [ -n "${CLIENT_CERT_FILE:-}" ]; then
+  export CLIENT_CERT_FILE="${CLIENT_CERT_FILE}"
```
This code looks like it sets the var to itself if it is not empty. Isn't that a no-op? What am i missing?

richscott replied:
Yes - this is wrong - it was a WIP idea that was expected to be filled with defaults, but is actually moot. Lines for CLIENT_CERT_FILE, CLIENT_KEY_FILE, and CLUSTER_CA_FILE here removed in commit c819e51.

richscott committed:
PR feedback review fixes.

Signed-off-by: Rich Scott <richscott@sent.com>

GeorgeJahad commented on scripts/dev-e2e.sh:
```diff
@@ -79,20 +79,38 @@ start-armada() {
   fi
 
-  echo  "Running 'make kind-all' to install and start Armada; this may take up to 6 minutes"
+  # Get IP address of first network interface that is not loopback or a K8S internal network interface
```
I could use a bigger comment, something like:
```
  # Get IP address of the Armada servers first network interface that
  # is not loopback or a K8S internal network interface and patch it
  # into the Kind config so that it binds to a valid address instead
  # of the hardcoded placeholder (192.168.12.135).
```

richscott replied:
👍🏼   This verbiage now adopted in commit 2bf3388.

richscott committed:
Derive k8s client encryption algorithm from client cert file.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Remove unused import.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on scripts/dev-e2e.sh:
```diff
@@ -79,20 +79,38 @@ start-armada() {
+    err "There was an error modifying $AOHOME/hack/kind-config.yaml"
+    exit 1
+  fi
+
```
Copilot is wrong here.  The manual page for sed (`man 1 sed`) on MacOS 15.7.5 has
```
     -I extension
             Edit files in-place, saving backups with the specified extension.  If a zero-length extension is given, no backup will be saved.  It is not recommended to give a zero-length
             extension when in-place editing files, as you risk corruption or partial content in situations where disk space is exhausted, etc.

             Note that in-place editing with -I still takes place in a single continuous line address space covering all files, although each file preserves its individuality instead of
             forming one output stream.  The line counter is never reset between files, address ranges can span file boundaries, and the “$” address matches only the last line of the
             last file.  (See Sed Addresses.) That can lead to unexpected results in many cases of in-place editing, where using -i is desired.
```
A casual test just now of doing:
```
sed -I .bak -e 's/the/foo/' testfile.txt
```
does produce a testfile.txt.bak containing the original contents, and testfile.txt now has all instances of "the" changed to "foo".

richscott committed:
Render error and exit if no external IP addrs found

Also use 'http', not 'https' for default Lookout URL.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on scripts/dev-e2e.sh:
```diff
@@ -79,20 +79,38 @@ start-armada() {
+  if ! sed "$sed_opt" -e "s/192.168.12.135/$external_ip/" "$AOHOME/hack/kind-config.yaml"; then
+    err "There was an error modifying $AOHOME/hack/kind-config.yaml"
+    exit 1
+  fi
```
fixed in commit 57320d7

richscott commented on scripts/init.sh:
```diff
@@ -104,6 +104,7 @@ export INCLUDE_PYTHON="${INCLUDE_PYTHON:-false}"
 export USE_KIND="${USE_KIND:-false}"
 export IMAGE_NAME="${IMAGE_NAME:-spark:armada}"
 export ARMADA_MASTER="${ARMADA_MASTER:-armada://localhost:30002}"
+export ARMADA_LOOKOUT_URL="${ARMADA_LOOKOUT_URL:-https://localhost:30000}"
```
fixed in commit 57320d7

richscott committed:
Merge branch 'rich/support-remote-server' of github.com:armadaproject/armada-spark into rich/support-remote-server

richscott committed:
Use current-context value from .kube/config to select cluster.

Also, declare `armadaCtl` as immutable from pattern-match on
ARMADA_MASTER setting.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Fix CLIENT_KEY_FILE misspelling

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on scripts/init.sh:
```diff
@@ -112,9 +113,34 @@ export ARMADA_EVENT_WATCHER_USE_TLS=${ARMADA_EVENT_WATCHER_USE_TLS:-false}
+  export CLIENT_CERT_FILE="${CLIENT_CERT_FILE}"
+fi
+if [ -n "${CLIENT_CERT_KEY:-}" ]; then
+  export CLIENT_CERT_KEY="${CLIENT_CERT_KEY}"
```
Fixed in commit 823ce95 - this was lurking in a Git-uncommitted change.

richscott committed:
Merge branch 'rich/support-remote-server' of github.com:armadaproject/armada-spark into rich/support-remote-server

richscott committed:
Note that Spark 3.3.4 support is future-planned in README.

PR review noted the incompatility of the supported versions in
README.md versus the currently commented-out versions (i.e. Spark
3.3.4) in the e2e.yaml Github Actions workflow.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on .github/workflows/e2e.yaml:
```diff
@@ -13,16 +13,16 @@ jobs:
-            java_version: "11"
+          # - scala_version: "2.12.15"
+          # spark_version: "3.3.4"
+          # java_version: "11"
           - scala_version: "2.12.18"
             spark_version: "3.5.5"
             java_version: "17"
 
-          - scala_version: "2.13.8"
-            spark_version: "3.3.4"
-            java_version: "11"
+          # - scala_version: "2.13.8"
+          # spark_version: "3.3.4"
+          # java_version: "11"
           - scala_version: "2.13.8"
             spark_version: "3.5.5"
             java_version: "17"
```
The Spark 3.3.4 versions have been moved to a "future planned" section in the README.md for now.

richscott committed:
Normalize/fix usage of ARMADA_MASTER setting.

Also remove dead var declaration in TestOrchestrator.scala.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on src/test/scala/org/apache/spark/deploy/armada/e2e/ArmadaSparkE2E.scala:
```diff
@@ -59,6 +61,11 @@ class ArmadaSparkE2E
 
     val props = loadProperties()
 
+    val armadaApiUrl = props.getProperty("armada.master", "localhost:30002")
```
Fixed in commit 2f81846

richscott commented on src/test/scala/org/apache/spark/deploy/armada/e2e/TestOrchestrator.scala:
```diff
@@ -605,12 +608,19 @@ class TestOrchestrator(
+    val driverClassPath = Seq(
+      ".",
+      "./target/armada-cluster-manager_2.13-1.0.0-SNAPSHOT-all.jar"
+    ).mkString(":")
```
Fixed in commit 2f81846

richscott commented on scripts/dev-e2e.sh:
```diff
@@ -138,45 +173,85 @@ init-cluster() {
-    -Darmadactl.path="$scripts/armadactl" 2>&1 | \
-    tee e2e-test.log
-
+    -Darmada.master="armada://$ARMADA_MASTER" \
```
Fixed in commit 2f81846

richscott commented on .github/workflows/build.yaml:
```diff
@@ -13,16 +13,16 @@ jobs:
-            java_version: "11"
+          # - scala_version: "2.12.15"
+          # spark_version: "3.3.4"
+          # java_version: "11"
           - scala_version: "2.12.18"
             spark_version: "3.5.5"
             java_version: "17"
 
-          - scala_version: "2.13.8"
-            spark_version: "3.3.4"
-            java_version: "11"
+          # - scala_version: "2.13.8"
+          # spark_version: "3.3.4"
+          # java_version: "11"
           - scala_version: "2.13.8"
             spark_version: "3.5.5"
             java_version: "17"
```
README.md has been updated in commit 4be3daf to mention that spark 3.3.4 is a future-planned supported version.

richscott committed:
Relax regex for matching docker image names

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on scripts/dev-e2e.sh:
```diff
@@ -102,6 +120,12 @@ init-cluster() {
+    err "INIT_CONTAINER_IMAGE is not defined. Please set it in $scripts/config.sh, for example:"
+    err "INIT_CONTAINER_IMAGE=busybox:latest"
+    exit 1
+  fi
```
The regex has been expanded to also allow `.` and `/` on the org/owner portion, in commit 22215e6 ; this matches the traditional `spark:armada` image name, as well as `some-org/spark:armada`, etc.

richscott committed:
Improve regex for matching container image tags.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
In init.sh, only set SPARK_LOCAL_IP if not already set.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on scripts/init.sh:
```diff
@@ -112,9 +113,34 @@ export ARMADA_EVENT_WATCHER_USE_TLS=${ARMADA_EVENT_WATCHER_USE_TLS:-false}
+# addresses, and use the first one that is not the loopback interface, or a
+# Docker/K8S virtual interface.
+export SPARK_LOCAL_IP=$(ifconfig -a | grep -w 'inet' | grep -Ev '(127\.0\.0\.1|\<172)' | awk '{print $2}' | sed -n '1p')
+if [ "$SPARK_LOCAL_IP" = "" ] ; then
```
Fixed in commit 83b8c9d

richscott commented on e2e/extract-kind-cert.sh:
```diff
@@ -0,0 +1,25 @@
+#  -  client.key: The private key associated with the certificate
+#  -  ca.crt: The CA certificate used by the Kubernetes API server (for verifying client and server certs)
+
+# Extract the client certificate
+kubectl config view --raw -o json | jq -r \
+  ".users[] | select(.name == \"${CONTEXT}\") | .user.[\"client-certificate-data\"]" | base64 -d > client.crt
+
+# Extract the client key
+kubectl config view --raw -o json | jq -r \
+  ".users[] | select(.name == \"${CONTEXT}\") | .user.[\"client-key-data\"]" | base64 -d > client.key
+
+# Extract the cluster CA certificate
+kubectl config view --raw -o json | jq -r \
+  ".clusters[] | select(.name == \"${CONTEXT}\") | .cluster.[\"certificate-authority-data\"]" | base64 -d > ca.crt
```
`base64 -d` is equivalent to `base64 -D` on macOS. The base64 manual page:
```
 -d, -D, --decode
            Decode incoming Base64 stream into binary data.
```

richscott committed:
MacOS compatibility fixes; remove unused var decl

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on scripts/dev-e2e.sh:
```diff
@@ -138,45 +173,85 @@ init-cluster() {
+    if [ -f "$spark_dockerfile" ]; then
+      sed -i -e 's|FROM openjdk:|FROM eclipse-temurin:|g' "$spark_dockerfile"
+      sed -i -E 's/^ARG java_image_tag=11-jre-slim$/ARG java_image_tag=11-jammy/' "$spark_dockerfile"
+    fi
```
Fixedd in commit 7e879dd

richscott replied:
Fixed in commit 7e879dd

richscott committed:
Fix default armada URL values in ArmadaClient test class

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Use more flexible heuristic for computing armadactl endpoint.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on src/test/scala/org/apache/spark/deploy/armada/e2e/ArmadaOperations.scala:
```diff
@@ -213,7 +213,20 @@ class ArmadaClient(armadaUrl: String = "localhost:30002") {
+          s"could not extract valid armadactl URL from armada URL ${armadaUrl}"
+        )
+    }
+
```
The suggested change has been implemented in commit 6c378b2

richscott committed:
Clarify language on configuring for external armada cluster.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Remove no-op settings of client cert file vars in init.sh

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Clarify comment about finding external network interface.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'master' into rich/support-remote-server

richscott committed:
Port-forwarding for port 7078 is not necessary with "host" network.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on src/test/scala/org/apache/spark/deploy/armada/e2e/TestOrchestrator.scala:
```diff
@@ -605,12 +608,19 @@ class TestOrchestrator(
-      "host"
+      "host",
+      "-p",
+      "7078:7078" // Driver port must be exposed for client deploy-mode
```
The `-p 7078:7078` port opening has been removed in commit 25af300 and it now uses only the "host"-based network setting.

richscott commented on src/test/scala/org/apache/spark/deploy/armada/e2e/TestOrchestrator.scala:
```diff
@@ -610,7 +611,9 @@ class TestOrchestrator(
-      "host"
+      "host",
+      "-p",
+      "7078:7078" // Driver port must be exposed for client deploy-mode
```
fixed in 25af300

richscott committed:
Scala formatter fix

Signed-off-by: Rich Scott <richscott@sent.com>

sudiptob2 commented on src/test/scala/org/apache/spark/deploy/armada/e2e/K8sClient.scala:
```diff
@@ -15,21 +15,82 @@
  */
 
-package org.apache.spark.deploy.armada.e2e
+package org.apache.spark.deploy.armada
```
Question: Is this package name correct now? Since the file is in /e2e/ shoundnt the package name be `org.apache.spark.deploy.armada.e2e.`

richscott replied:
Puzzling that this compiled and ran okay as it is - anyway, the package decl and refs for K8sClient are now normalized as org.apache.spark.deploy.armada.e2e in commit 98f3c51.

sudiptob2 commented on src/test/scala/org/apache/spark/deploy/armada/e2e/K8sClient.scala:
```diff
@@ -15,21 +15,82 @@
-import io.fabric8.kubernetes.client.{DefaultKubernetesClient, KubernetesClient}
+import io.fabric8.kubernetes.client.{
+  ConfigBuilder,
+  DefaultKubernetesClient,
```
DefaultKubernetesClient seems unused

richscott replied:
Fixed in commit 448a09a.

sudiptob2 commented on src/test/scala/org/apache/spark/deploy/armada/e2e/K8sClient.scala:
```diff
@@ -15,21 +15,82 @@
+  val data      = yaml.load[java.util.Map[String, Object]](new FileReader(s"$home/.kube/config"))
+  var k8sApiURL = "no-K8S-server-found"
+
+  val activeCluster = data.get("current-context").asInstanceOf[String]
```
It seems that, in **kind**, the context name and the cluster name are the same so this works. 
But, for other setups such as **minikube** or **k3s**, the cluster name may differ from the current context name. Can we confirm that? If needed, I guess it is fine to limit support to **kind** for now and create a ticket to track this limitation.

richscott replied:
After some more research, it seems that this heuristic needs to be expanded to a 2-step search, so it works with all k8s implementations:
1. get the value of `current-context`
2. iterate through the list of the `contexts` array; if an entry in that list has its "name" value matching the value found in step 1, save the value associated with the "cluster" key in that entry.
3. using the specific cluster name found in step 2, iterate through "clusters" list block, and use the value of the "server" key if the "name" value equals the cluster name found in step 2.

richscott replied:
This logic has been replaced with usage of the fabric8 `Config.autoConfigure()` method, which is much shorter, and appears to be reliable in multiple configurations/k8s-providers, in commit 7ff8595.

sudiptob2 commented on e2e/extract-kind-cert.sh:
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+CONTEXT="kind-armada"
```
Should this always be `kind-armada`.

sudiptob2 reviewed (commented):
Thanks for addressing the previous reviews. Looking good. I gave it a re-review, and here are a few small comments I have.

sudiptob2 commented on README.md:
```diff
@@ -79,6 +84,90 @@ We recommend [kind](https://kind.sigs.k8s.io/) for local testing. If you are usi
+
+- Run the armada-operator setup commands (usually `make kind-all`) to create and start your Armada instance.
+
+- Copy the `$HOME/.kube/config` and `$HOME/.armadctl.yaml` (that Armada Operator will generate) from the Armada
```
Agree, its a typo

richscott replied:
fixed in commit 1558f8e

richscott committed:
Typo fix in README.md.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on scripts/init.sh:
```diff
@@ -112,9 +113,29 @@ export ARMADA_EVENT_WATCHER_USE_TLS=${ARMADA_EVENT_WATCHER_USE_TLS:-false}
+# addresses, and use the first one that is not the loopback interface, or a
+# Docker/K8S virtual interface.
+if [ -z "${SPARK_LOCAL_IP:-}" ]; then
+  SPARK_LOCAL_IP=$(ifconfig -a | grep -w 'inet' | grep -Ev '(127\.0\.0\.1|\<172)' | awk '{print $2}' | sed -n '1p')
```
This code is already precarious enough. Will change it when an observed failure occurs in practice.

richscott commented on e2e/extract-kind-cert.sh:
```diff
@@ -0,0 +1,25 @@
+
+CONTEXT="kind-armada"
+
+E2E_DIR=$(realpath "$0" | xargs dirname)
```
`/bin/realpath` is installed by default on all reasonably contemporary macOS releases.

richscott committed:
Normalize org.apache.spark.deploy.armada.e2e package for K8sClient

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Remove unused import

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Use fabric8 Config package to parser kube config.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on src/test/scala/org/apache/spark/deploy/armada/e2e/K8sClient.scala:
```diff
@@ -15,21 +15,82 @@
+      val cluster = entry.get("cluster").asInstanceOf[java.util.Map[String, Object]]
+      k8sApiURL = cluster.get("server").toString()
+    }
+  }
```
In commit 7ff8595 this code has been replaced with a use of the fabric8 `Config.autoConfigure()` method, which appears to reliably derive the current context and cluster info.

richscott commented on src/test/scala/org/apache/spark/deploy/armada/e2e/K8sClient.scala:
```diff
@@ -15,21 +15,82 @@
+
+    val cf   = CertificateFactory.getInstance("X.509")
+    val cert = cf.generateCertificate(new FileInputStream(clientCertFile))
+    algo = cert.getPublicKey.getAlgorithm // "RSA" or "EC"
```
This code has been subsumed by a replacement that uses the fabric8 `Config.autoConfigure()` method, in commit 7ff8595.

richscott committed:
Replace hard-wired k8s context with dynamic value.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on scripts/dev-e2e.sh:
```diff
@@ -7,7 +7,7 @@ source "$scripts"/init.sh
 STATUSFILE="$(mktemp)"
 AOREPO='https://github.com/armadaproject/armada-operator.git'
-AOHOME="$scripts/../../armada-operator"
+AOHOME=$(realpath "$scripts/../../armada-operator")
```
`realpath` is available on at least the last two major MacOS releases, and any basic Ubuntu distribution, since it's part of GNU coreutils.

richscott committed:
Improve parsing of Armada Master URL value

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on src/main/scala/org/apache/spark/deploy/armada/submit/ArmadaUtils.scala:
```diff
@@ -28,11 +28,7 @@ object ArmadaUtils {
-      "armada://"
-    }
+    val startString = "armada://"
     Some(masterUrl)
```
commit 31a5355

richscott committed:
Refine sed usage, per Copilot suggestion.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on scripts/dev-e2e.sh:
```diff
@@ -16,6 +16,16 @@ NC='\033[0m'
+
+sed_backup='-I .bak'
+else
+sed_backup='-i.bak'
```
Using the suggested sed command array has been now implemented in commit 73c730d.

sudiptob2 commented on scripts/dev-e2e.sh:
```diff
@@ -138,45 +207,84 @@ init-cluster() {
+  ./build/mvn --batch-mode package -pl examples
+  ./build/mvn --batch-mode package -Pkubernetes -Pscala-$SCALA_BIN_VERSION -pl assembly
+  cd ..
+
```
The test runs `spark-submit` via Docker (see [TestOrchestrator.scala](https://github.com/armadaproject/armada-spark/blob/c3b27339dd2c60770ceff49e431945ac9fd5b963/src/test/scala/org/apache/spark/deploy/armada/e2e/TestOrchestrator.scala#L608-L624)). The required JARs are already built and baked into the image by `createImage.sh`.

https://github.com/armadaproject/armada-spark/blob/c3b27339dd2c60770ceff49e431945ac9fd5b963/scripts/createImage.sh#L46-L52

Given that, rebuilding or recreating them during `run-test` appears redundant. It also adds overhead when running e2e tests repeatedly. Let me know if I am missing a scenario where this step is still required?

richscott replied:
Thanks - this redundant chunk removed in commit 6f584cb.

GeorgeJahad commented on src/test/scala/org/apache/spark/deploy/armada/e2e/ArmadaOperations.scala:
```diff
@@ -213,7 +213,16 @@ class ArmadaClient(armadaUrl: String = "localhost:30002") {
+    // armadactl command expects the server address to be of the form
+    // <hostname-or-IP>:<port> with no pseudo-protocol prefix
+    val armadactlUrl =
+      if (armadaUrl.startsWith("armada://")) {
```
stripPrefix is a no-op when the prefix isn't present, so the if is unnecessary:
```
  val armadactlUrl = armadaUrl.stripPrefix("armada://")    
```

richscott replied:
Good catch - fixed in commit 6f584cb.

richscott committed:
Remove redundant compilation of Spark source in E2E.

Also remove superfluous conditional for using String.stripPrefix().

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'master' into rich/support-remote-server

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
PR review fixes

- don't put error exit in subshell in extract-kind-cert.sh
- more portable grep pattern for filtering out local/k8s addresses
- fetch properties for TLS cert file env vars
- remove some dead imports and vars in K8sClient.scala
- use suggested comment on SPARK_LOCAL_IP explanation

Signed-off-by: Rich Scott <richscott@sent.com>

GeorgeJahad reviewed (approved):
lgtm, thanks @richscott !

richscott commented on src/test/scala/org/apache/spark/deploy/armada/e2e/ArmadaSparkE2E.scala:
```diff
@@ -59,6 +61,11 @@ class ArmadaSparkE2E
 
     val props = loadProperties()
 
+    val armadaApiUrl = props.getProperty("armada.master", "armada://localhost:30002")
+    armadaClient = new ArmadaClient(armadaApiUrl)
+    k8sClient = new K8sClient(props)
+    orch = new TestOrchestrator(armadaClient, k8sClient)
```
Fixed in commit e45ba02

richscott commented on src/test/scala/org/apache/spark/deploy/armada/e2e/K8sClient.scala:
```diff
@@ -18,18 +18,72 @@
+  val kubeCfg   = KubeConfig.autoConfigure(null)
+  val k8sApiURL = kubeCfg.getMasterUrl()
+
+  val clientCertFile: String = props.getProperty("client_cert_file", "")
+  val clientKeyFile: String  = props.getProperty("client_key_file", "")
+  val clusterCaFile: String  = props.getProperty("cluster_ca_file", "")
```
fixed in commit e45ba02

richscott commented on src/test/scala/org/apache/spark/deploy/armada/e2e/K8sClient.scala:
```diff
@@ -18,18 +18,72 @@
+class K8sClient(props: Properties) {
+  val armadaMaster: String = props.getProperty("armada.master")
+  val pattern              = """armada://([^:]+):.*""".r
+
```
Removed in commit e45ba02

richscott commented on scripts/init.sh:
```diff
@@ -112,9 +113,29 @@ export ARMADA_EVENT_WATCHER_USE_TLS=${ARMADA_EVENT_WATCHER_USE_TLS:-false}
+# addresses, and use the first one that is not the loopback interface, or a
+# Docker/K8S virtual interface.
+if [ -z "${SPARK_LOCAL_IP:-}" ]; then
+  SPARK_LOCAL_IP=$(ifconfig -a | grep -w 'inet' | grep -Ev '(127\.0\.0\.1|\<172)' | awk '{print $2}' | sed -n '1p')
```
Suggested fix in commit e45ba02

richscott commented on e2e/extract-kind-cert.sh:
```diff
@@ -0,0 +1,25 @@
+
+E2E_DIR=$(realpath "$0" | xargs dirname)
+
+cd "$E2E_DIR" || (echo "Error: could not cd to $E2E_DIR"; exit 1)
```
Suggested fix in commit e45ba02.

richscott commented on src/test/scala/org/apache/spark/deploy/armada/e2e/TemplateFileServer.scala:
```diff
@@ -30,9 +30,16 @@ import com.sun.net.httpserver.{HttpExchange, HttpServer}
+  // Kind pods (which use CoreDNS and don't inherit Docker's host alias). SPARK_LOCAL_IP
+  // is the host machine's LAN IP, reachable from both Docker containers (via --network host)
+  // and Kind pods (confirmed: executor pods already connect to this IP in client mode).
+  // On Linux, 172.18.0.1 is the Docker bridge gateway, accessible from both.
```
Suggested text change in commit e45ba02.

richscott commented on README.md:
````diff
@@ -79,6 +84,90 @@ We recommend [kind](https://kind.sigs.k8s.io/) for local testing. If you are usi
+Armada server, as well as running `armadactl get queues`.
+
+- Verify the functionality of your setup by editing `scripts/config.sh` and changing the following line:
+```
+ARMADA_MASTER=armada://192.168.12.135:30002
+```
+to the IP address or hostname of your Armada server. You should not need to change the port number.
+
+Also, set the location of the three TLS certificate files by adding/setting:
+```
+export CLIENT_CERT_FILE=e2e/client.crt
+export CLIENT_KEY_FILE=e2e/client.key
+export CLUSTER_CA_FILE=e2e/ca.crt
+```
+
+-  You should be able to now verify the armada-spark configuration by running the E2E tests:
+```
+$ ./scripts/dev-e2e.sh
+```
+This will save its output to `e2e-test.log` for further debugging.
````
I am leaving the code blocks un-indented as the other pre-existing code blocks already are.

richscott committed:
Merge branch 'master' into rich/support-remote-server

sudiptob2 reviewed (approved):
Thanks for addressing the reviews. lgtm
